### ISSUE TITLE
implement custom loss for multi-output training that ignores unknown outputs

### ISSUE THREAD

**Issue by timodonnell** [hammerlab/pepnet#7]
If an output value for a data point is NaN, the corresponding output shouldn't impact the loss for that data point. Necessary for multi-output mass-spec + affinity predictors since for most points we know one output value but not the other.

@iskandr I'm down to implement this but just want to make sure you haven't already so we don't duplicate work?

**Comment by iskandr** [comment]
I have an implementation using negative values as the sentinel (instead of NaN) because handling NaN's via Keras's backend seemed tricky. My old solution was to use an explicit `Input` to mask out each output but with multiple outputs this made for ugly/complex models (since each output must be wired to its input via a custom loss and the masks must be passed in by pepnet). Thinking about going back to that, but curious to see if you find a more elegant way to do things.

**Comment by iskandr** [comment]
Specifically, I didn't find anything like `keras.backend.isnan` but we need that to properly mask out the losses.

**Comment by iskandr** [comment]
For reference, my old masked loss was like this: 
https://github.com/matthiasplappert/keras-rl/blob/master/rl/agents/dqn.py#L183

(currently lives in a notebook on my laptop)

**Comment by timodonnell** [comment]
Thanks @iskandr . Also possibly relevant (seems he got it working on theano backend?): https://github.com/fchollet/keras/issues/1628

**Comment by iskandr** [comment]
Yeah, I was thinking about using T.isnan but I didn't like reaching underneath the keras backend abstraction. But maybe that's the cleanest solution.

**Comment by timodonnell** [comment]
This seems to work for me, although it's theano-specific

```python
import keras.backend as K
import theano.tensor
def custom_loss(y_true, y_pred):
    squared = K.square(y_pred - y_true)
    loss = K.sum(
        K.switch(theano.tensor.isnan(y_true), 0.0, squared),
        axis=-1)
    return loss
```

I'll make a PR if this seems good to you @iskandr

**Comment by iskandr** [comment]
Fixed by https://github.com/hammerlab/pepnet/pull/6
